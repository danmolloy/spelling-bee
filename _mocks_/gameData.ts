import { faker } from "@faker-js/faker"

//const centerLetter = faker.string.alpha();
//const outerLetters = faker.string.alpha(6).split("")
const expiration = Date.now() + 10000
const answers = faker.word.words(25).toLowerCase().split("").filter(i => i !== " ")
const lettersArr = Array.from(new Set(answers)).slice(0, 7)

const validLetters = Array.from(new Set(faker.string.alpha(12).split(""))).slice(0, 7)

export const mockData = {
  expiration: expiration,
  displayWeekday:"Sunday",
  displayDate: "October 13, 1986",
  printDate:"2022-04-10",
  centerLetter: lettersArr[0],
  outerLetters: lettersArr.slice(1, 7),
  validLetters: lettersArr,
  pangrams: [lettersArr.join("")],
  answers: [lettersArr.join(""), lettersArr.slice(1, 6).join(""), ...faker.word.words(25).split(" ")],
  id: faker.string.uuid(),
  freeExpiration:"0",
  editor:"Sam Ezersky"
}