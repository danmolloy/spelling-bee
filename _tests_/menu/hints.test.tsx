import "@testing-library/jest-dom"
import { screen, render, act, fireEvent } from "@testing-library/react"
import Hints, { HintsProps } from "../../components/menu/hints"
import { mockData } from "../../_mocks_/gameData"

jest.spyOn(Storage.prototype, 'setItem')
Storage.prototype.setItem = jest.fn()

const mockProps: HintsProps = {
  setShowMenuItem: jest.fn(),
  answers: mockData.answers,
  pangrams: ['lorem'],
  setRevealAnswers: jest.fn(),
  revealAnswers: false
}

describe("<Hints />", () => {
  beforeEach(() => {
    render(<Hints {...mockProps} />)
  })
  it("hints-div is in the document", () => {
    const hintsDiv = screen.getByTestId("hints-div")
    expect(hintsDiv).toBeInTheDocument()
  })
  it("'Hints' title is in the document", () => {
    const hintsTitle = screen.getByText("Hints")
    expect(hintsTitle).toBeInTheDocument()
  })
  it("menu-icon is in the document and calls setShowMenuItem(null) on click", () => {
    const menuIcon = screen.getByTestId("menu-icon")
    expect(menuIcon).toBeInTheDocument()
  })
  it("states how many answers are in puzzle", () => {
    const hintsDiv = screen.getByTestId("hints-div")
    expect(hintsDiv.textContent).toMatch(`The current puzzle has ${mockProps.answers.length} words.`)
  })
  it("states number of pangrams in the singular", () => {
    const hintsDiv = screen.getByTestId("hints-div")
    expect(hintsDiv.textContent).toMatch(`There is 1 pangram.`)

  })
  it("states number of words for each length", () => {
    const hintsDiv = screen.getByTestId("hints-div")
    expect(hintsDiv.textContent).toMatch(`Four letters: ${mockProps.answers.filter(i => i.length === 4).length}`)
    expect(hintsDiv.textContent).toMatch(`Five letters: ${mockProps.answers.filter(i => i.length === 5).length}`)
    expect(hintsDiv.textContent).toMatch(`Six letters: ${mockProps.answers.filter(i => i.length === 6).length}`)
    expect(hintsDiv.textContent).toMatch(`Seven letters: ${mockProps.answers.filter(i => i.length === 7).length}`)
    expect(hintsDiv.textContent).toMatch(`Eight or more letters: ${mockProps.answers.filter(i => i.length >= 8).length}`)
  })
  it("reveal btn is in the document and calls setRevealAnswers & setShowMenuItem on click", () => {
    const revealBtn = screen.getByText("Reveal")
    expect(revealBtn).toBeInTheDocument()
    act(() => {
      fireEvent.click(revealBtn)
    })
    expect(mockProps.setRevealAnswers).toHaveBeenCalled()
    expect(mockProps.setShowMenuItem).toHaveBeenCalledWith(null)
    expect(localStorage.setItem).toHaveBeenCalledWith("revealed", "true")

  })
})

describe("<Hints />", () => {
  const mockProps: HintsProps = {
    setShowMenuItem: jest.fn(),
    answers: ["four", "funf", "sixsix", "seventy", "eighty"],
    pangrams: ["lorem", "ipsum"],
    setRevealAnswers: jest.fn(),
    revealAnswers: false
  }
  beforeEach(() => {
    render(<Hints {...mockProps} />)
  })
  it("states number of pangrams in the plural", () => {
    const hintsDiv = screen.getByTestId("hints-div")
    expect(hintsDiv.textContent).toMatch(`There are 2 pangrams.`)
  })
  it("matches snapshot", () => {
    const hintsDiv = screen.getByTestId("hints-div")
    expect(hintsDiv).toMatchSnapshot()
  })
})