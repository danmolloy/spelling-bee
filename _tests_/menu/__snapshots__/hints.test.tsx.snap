// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hints /> matches snapshot 1`] = `
<div
  data-testid="hints-div"
>
  <div
    class="flex flex-row w-full justify-between items-center"
  >
    <h1
      class="text-2xl font-bold"
    >
      Hints
    </h1>
    <button
      class="menu-icon"
      data-testid="menu-icon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
        />
      </svg>
    </button>
  </div>
  <div
    class="p-2"
  >
    <p
      class="text-lg font-thin"
    >
      Words: 
      5
    </p>
    <p
      class="text-lg font-thin"
    >
      Pangrams: 
      2
    </p>
    <p
      class="text-lg font-thin"
    >
      Points: 
      21
    </p>
  </div>
  <div
    class="flex flex-col items-center py-2"
  >
    <table
      class="w-2/3 md:w-1/2"
      data-testid="help-table"
    >
      <thead
        data-testid="table-head"
      >
        <tr
          class="font-bold"
        >
          <td
            data-testid="letter-col"
          />
          <td>
            <p
              class="text-center"
            >
              4
            </p>
          </td>
          <td>
            <p
              class="text-center"
            >
              6
            </p>
          </td>
          <td>
            <p
              class="text-center"
            >
              7
            </p>
          </td>
          <td>
            <p
              class="text-center"
            >
              ∑
            </p>
          </td>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          data-testid="q-row"
        >
          <td
            class="font-bold"
          >
            Q
          </td>
          <td
            class=""
            data-testid="4-q"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="6-q"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="7-q"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            data-testid="q-total"
          >
            <p
              class="text-center"
            >
              0
            </p>
          </td>
        </tr>
        <tr
          class=""
          data-testid="a-row"
        >
          <td
            class="font-bold"
          >
            A
          </td>
          <td
            class=""
            data-testid="4-a"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="6-a"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="7-a"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            data-testid="a-total"
          >
            <p
              class="text-center"
            >
              0
            </p>
          </td>
        </tr>
        <tr
          class=""
          data-testid="b-row"
        >
          <td
            class="font-bold"
          >
            B
          </td>
          <td
            class=""
            data-testid="4-b"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="6-b"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="7-b"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            data-testid="b-total"
          >
            <p
              class="text-center"
            >
              0
            </p>
          </td>
        </tr>
        <tr
          class=""
          data-testid="c-row"
        >
          <td
            class="font-bold"
          >
            C
          </td>
          <td
            class=""
            data-testid="4-c"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="6-c"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="7-c"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            data-testid="c-total"
          >
            <p
              class="text-center"
            >
              0
            </p>
          </td>
        </tr>
        <tr
          class=""
          data-testid="d-row"
        >
          <td
            class="font-bold"
          >
            D
          </td>
          <td
            class=""
            data-testid="4-d"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="6-d"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="7-d"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            data-testid="d-total"
          >
            <p
              class="text-center"
            >
              0
            </p>
          </td>
        </tr>
        <tr
          class=""
          data-testid="e-row"
        >
          <td
            class="font-bold"
          >
            E
          </td>
          <td
            class=""
            data-testid="4-e"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="6-e"
          >
            <p
              class="text-center"
            >
              1
            </p>
          </td>
          <td
            class=""
            data-testid="7-e"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            data-testid="e-total"
          >
            <p
              class="text-center"
            >
              1
            </p>
          </td>
        </tr>
        <tr
          class=""
          data-testid="f-row"
        >
          <td
            class="font-bold"
          >
            F
          </td>
          <td
            class=""
            data-testid="4-f"
          >
            <p
              class="text-center"
            >
              2
            </p>
          </td>
          <td
            class=""
            data-testid="6-f"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            class=""
            data-testid="7-f"
          >
            <p
              class="text-center"
            >
              -
            </p>
          </td>
          <td
            data-testid="f-total"
          >
            <p
              class="text-center"
            >
              2
            </p>
          </td>
        </tr>
        <tr
          data-testid="total-row"
        >
          <td
            class="font-bold"
          >
            ∑
          </td>
          <td
            data-testid="4-total"
          >
            <p
              class="text-center"
            >
              2
            </p>
          </td>
          <td
            data-testid="6-total"
          >
            <p
              class="text-center"
            >
              2
            </p>
          </td>
          <td
            data-testid="7-total"
          >
            <p
              class="text-center"
            >
              1
            </p>
          </td>
          <td
            class="text-center"
            data-testid="total-words"
          >
            5
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="p-2 my-2"
    data-testid="two-letter-list"
  >
    <p
      class="font-bold"
    >
      Two letter list:
    </p>
    <div
      class="flex flex-row "
      data-testid="e-row"
    >
      <p
        class="pr-1"
      >
        EI-1
      </p>
    </div>
    <div
      class="flex flex-row "
      data-testid="f-row"
    >
      <p
        class="pr-1"
      >
        FO-1
      </p>
      <p
        class="pr-1"
      >
        FU-1
      </p>
    </div>
    <div
      class="flex flex-row "
      data-testid="s-row"
    >
      <p
        class="pr-1"
      >
        SE-1
      </p>
      <p
        class="pr-1"
      >
        SI-1
      </p>
    </div>
  </div>
  <hr />
  <div
    class="my-2"
  >
    <h2
      class="text-xl font-medium py-1 my-2"
    >
      Reveal Answers
    </h2>
    <div
      class="how-to-section flex flex-col"
    >
      <p
        class="font-semithin"
      >
        Reveal the answer list for the current game.
      </p>
      <p
        class="font-semithin"
      >
        This will end your gameplay until the next game.
      </p>
      <button
        class="self-center single-btn border-0 m-1 text-bold bg-yellow-300 hover:bg-yellow-200 active:bg-black active:text-yellow-300"
      >
        Reveal
      </button>
    </div>
  </div>
</div>
`;
